### PR TITLE
Basin Dao and Tests

### PR DESCRIPTION
**Pull request by zack-rma**
Basin Dao to match new Basin object

### CODE REVIEW

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        cleanUpRoutine();
+    }
+
+    @Test
+    void testDeleteBasin() throws Exception {
+        final String DELETE_ACTION = "NULL";
+
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = dslContext(c);
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: use getDslContext(c, OFFICE_ID);

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            }
+        });
+        cleanUpRoutine();
+    }
+
+    @Test
+    void testStoreBasin() throws Exception {
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = dslContext(c);
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: use getDslContext(c, OFFICE_ID);

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                locationsCreated.add(location);
+            }
+        }, "cwms_20");
+    }
+
+    @Test
+    void testGetBasin() throws Exception {
+        CwmsDatabaseContainer<?> db = CwmsDataApiSetupCallback.getDatabaseLink();
+        db.connection(c -> {
+            DSLContext ctx = dslContext(c);
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: use getDslContext(c, OFFICE_ID);

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                    }
+                },"cwms_20");
+                it.remove();
+            } catch(SQLException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    private static void createTestLocation(String locationName, String kind) throws SQLException{
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java:131]: I would use the existing locationDao to do this. I also think you need to specify a Nation and State ("CO") for this.

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        CwmsDbBasinJooq basinJooq = new CwmsDbBasinJooq();
+
+        basin.validate();
+
+        try {
+            String basinId = basin.getBasinId().getName();
+            String officeId = basin.getBasinId().getOfficeId();
+            String parentBasinId = basin.getParentBasinId() != null ? basin.getParentBasinId().getName() : null;
+            Double sortOrder = basin.getSortOrder();
+            String primaryStreamId = basin.getPrimaryStreamId().getName();
```
Comment (rma-bryson) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: I believe primaryStreamID can be null. Null check this (and anywhere else something similar is done)

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+class BasinDaoTest extends DataApiTestIT {
+    private static final String OFFICE_ID = TestAccounts.KeyUser.SWT_NORMAL.getOperatingOffice();
+    private static final String UNIT_SYSTEM = "SI";
+    private static final ArrayList<Location> locationsCreated = new ArrayList<>();
+    private static final Basin parentBasin = new Basin.Builder()
+            .withBasinId(new LocationIdentifier.Builder()
+                    .withName("TEST_PARENT_BASIN")
+                    .withOfficeId(OFFICE_ID)
+                    .build())
+            .withAreaUnit(UNIT_SYSTEM)
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: I believe this is the actual area unit "m2" here not the unit system

Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest push

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java (hunk 5 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: Addressed with a null check for now until Streams are fully implemented

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 4 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java:131]: Addressed in latest push

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 3 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest push

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 2 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest push

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 1 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest push

**Review by rma-bryson — APPROVED** [review]

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    static void setup() {
+        try {
+            testStream = new Stream.Builder()
+                    .withStartsDownstream(true)
+                    .withId(new CwmsId.Builder()
+                            .withName("TEST_PAR_PRIM_STREAM")
+                            .withOfficeId(OFFICE_ID)
+                            .build())
+                    .withComment("TESTING TESTING")
+                    .withSlopeUnits("m/km")
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: slope units are "%"

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+    private static void assertSame(Basin first, Basin second){
+        assertAll(
+                () -> {
+                    if (first.getBasinId() != null) {
+                        assertEquals(first.getBasinId().getName(), second.getBasinId().getName(), "Basin IDs are not equal");
+                        assertEquals(first.getBasinId().getOfficeId(), second.getBasinId().getOfficeId(), "Basin Offices are not equal");
+                    }
+                },
+                () -> {
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: for comparing the stream Ids use CwmsId.assertSame(...)

**Review by rma-bryson — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+                        .withName("TEST_PAR_PRIM_STREAM")
+                        .withOfficeId(OFFICE_ID)
+                        .build())
+                .build();
+    }
+
+    private static void assertSame(Basin first, Basin second){
+        assertAll(
+                () -> {
+                    if (first.getBasinId() != null) {
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Also use BasinTest.assertSame here to ensure we don't miss something

Comment (rma-bryson) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: I would just remove this assertSame and use the one in BasinTest. Make it public static, and also update the BasinTest's assertSame to use the assertSame in CwmsIdTest

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 8 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: See above comment about removing this assertSame altogether and instead use one defined in BasinTest

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 7 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest commit

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 8 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest commit

File: cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java (hunk 9 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dao/basin/BasinDaoTest.java]: Fixed in latest commit

**Review by rma-bryson — APPROVED** [review]

**Comment by adamkorynta** [comment]
Squash up the commits before ready to merge

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+public class BasinDao extends JooqDao<Basin> {
+
+    public BasinDao(DSLContext dsl) {
+        super(dsl);
+    }
+
+    public List<Basin> getAllBasins(String unitSystem, String officeId) throws SQLException {
+        List<Basin> retVal = new ArrayList<>();
+        CwmsDbBasinJooq basinJooq = new CwmsDbBasinJooq();
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: remove CwmsDbBasinJooq references

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        String areaUnitIn = UnitSystem.EN.value().equals(unitSystem)
+                ? Unit.SQUARE_MILES.getValue() : Unit.SQUARE_KILOMETERS.getValue();
+        try {
+            connection(dsl, c -> {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: use connectionResult for getOne/getMulti

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/test/java/cwms/cda/data/dto/basin/BasinTest.java [hunk 12]
```diff
@@ -121,10 +128,9 @@ private Basin buildTestBasin(CwmsId basinId, boolean parentBasinIdNull) {
         }
     }
 
-    private static void assertSame(Basin first, Basin second){
+    public static void assertSame(Basin first, Basin second){
         assertAll(
-                () -> assertEquals(first.getBasinId().getName(), second.getBasinId().getName(), "Basin IDs are not equal"),
-                () -> assertEquals(first.getBasinId().getOfficeId(), second.getBasinId().getOfficeId(), "Basin Office IDs are not equal"),
+                () -> CwmsIdTest.assertSame(first.getBasinId(), second.getBasinId()),
                 () -> assertEquals(first.getPrimaryStreamId().getName(), second.getPrimaryStreamId().getName(), "Primary Stream IDs are not equal"),
```
Comment (rma-bryson) [inline cwms-data-api/src/test/java/cwms/cda/data/dto/basin/BasinTest.java]: This should also use CwmsIdTest assertSame to compare the primary stream ids. Same with the parent basin ids below.

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java (hunk 10 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: Fixed in latest commit

File: cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java (hunk 11 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java]: fixed in latest commit

File: cwms-data-api/src/test/java/cwms/cda/data/dto/basin/BasinTest.java (hunk 12 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/data/dto/basin/BasinTest.java]: Addressed in commit

**Review by rma-bryson — APPROVED** [review]

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**